### DISCUSSION TITLE
How to upload from request.stream to MinIO

### DISCUSSION THREAD

**Discussion by zapfdk** [fastapi/fastapi#8403] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
from minio import Minio

minio_client = Minio(HOST, access_key=ACCESS_KEY, SECRET_KEY)

@app.post("/upload")
async def (request: Request):
    async for chunk in request.stream():
        result = minio_client.put_object(
            "files",
            "object",
            data=chunk,
            length=len(chunk),
    )
```

### Description

Hello :)
I would like to upload POSTed files directly to MinIO object storage.
I want to stream files directly from the body to  Minio, to avoid multiple copies of the file on the server.
Also I would like to avoid the UploadFile API, because I would like to avoid the overhead for multipart-parsing for bigger files.

I just can't wrap my head around how to get the chunks from the async generator from the request.stream to a read() interface as the Minio Client expects.
The example above doesn't work how I want, as the call will just simply replace previous put operations.
Also I just can't put request.stream in, as the minio_client can't handle the generator. It expects an object with a read() interface.

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

3.8.15

**Comment by frankie567** [comment]
I'm not very familiar with MinIO, but if `put_object` expects a `read`-like interface, the solution is to use the `UploadFile` and `File` dependency:

```py
from fastapi import FastAPI, File, UploadFile

@app.post("/upload")
async def (file: UploadFile = File(...)):
    result = minio_client.put_object(
        "files",
        "object",
        data=file,
        content_type=file.content_type
    )
```

The good thing with `UploadFile` is that it'll store the file you're uploading in memory up to a threshold. It means that it'll work even for very large files as they will be stored temporarily on disk.

Ref: https://fastapi.tiangolo.com/tutorial/request-files/#file-parameters-with-uploadfile

**Comment by zapfdk** [comment]
> I'm not very familiar with MinIO, but if `put_object` expects a `read`-like interface, the solution is to use the `UploadFile` and `File` dependency:

Yes, that was my first approach, too.
This automatically expects a `multipart/form-data` type though, which adds a significant overhead for parsing the multipart data for larger files.

What I did in the end:
I used `request.stream()` directly, and built a wrapper around it, using a buffer (in a Queue), exposing a `read`-like interface.
Works pretty well, although a bit more complicated.

**Reply by winstxnhdw** [reply]
`request.stream()` is of type `AsyncGenerator[bytes, None]`. I am curious how you handled this since `read()` has to be synchronous.

**Comment by frankie567** [comment]
I see! Another solution I already used is to actually directly upload to the S3 server, without passing through an intermediate API.

From your backend, It's possible to generate a presigned (authenticated) URL on your S3 bucket, which you return to your client browser. This URL allows the browser to directly upload the data to the bucket.

**Comment by wey-gu** [comment, marked as answer]
For anyone looking for a solution, here is mine that actually works:

```python
...
    file_content = file.file
    file_content.seek(0, os.SEEK_END)
    file_size = file_content.tell()
    file_content.seek(0)
    result = minio_client.put_object(
            "files",
            "object",
            data=file_content,
            content_type=file.content_type,
            length=file_size
        )
```
